### PR TITLE
Hoist constants so that they are not repeatedly rebuilt.

### PR DESCRIPTION
**Pull request by rixner**
This gets a pretty decent speedup on programs that have constants inside of loops by only turning those constants into python objects once at compile time instead of every time the constant is evaluated at run time.

The only potentially "controversial" change here is that I also turn tuples into constants if all their constituent parts are also constants.  This means that two identical tuples are the same object.  This is not the behavior of CPython where they become two independent objects, though I am not convinced that the language specification requires that.

This causes it to fail t429.py.  I commented out the one line in that file that tests that ```t1 is t2``` evaluates to ```False``` and ```t1 is not t2``` evaluates to ```True``` for identical tuples.

### CODE REVIEW

**Comment by rixner** [comment]
@meredydd the performance contribution is effectively impossible to quantify, as it completely depends on how many repeated tuple constructions there are.  Pretty much all programs have string/numeric constants and get some benefit from that part of it.  If those constants are within loops, there there is a significant benefit.  The same applies for tuples, but I'm not willing to say that "all/most" programs construct constant tuples inside loops...

It does not hurt performance, and it sometimes helps quite a lot, so I think it's worth it (and it will remain in my code regardless).  The *only* way you can tell is if you are using ```id``` or ```is``` on those tuples, which is presumably really, really rare, and furthermore I do not believe CPython specifically specifies what should happen in these cases, it just behaves the way it does because of implementation details not language specification.

In my opinion, not matching CPython here is completely harmless.  There are many places where we don't match CPython.  This is for a well-considered reason, not lack of completeness.

I could always add a flag, so you could specify whether you want this in the skulpt configuration.  I don't like the idea of proliferating too many flags, though, as that significantly complicates regression testing, and it will probably just get left out of such testing.

File: src/compile.js [hunk 1]
```diff
@@ -355,11 +390,31 @@ Compiler.prototype.ctuplelistorset = function(e, data, tuporlist) {
...
     else if (e.ctx === Load || tuporlist === "set") { //because set's can't be assigned to.
-        items = [];
-        for (i = 0; i < e.elts.length; ++i) {
-            items.push(this._gr("elem", this.vexpr(e.elts[i])));
+        if (tuporlist === "tuple") {
+            allconsts = true;
+            items = [];
+            for (i = 0; i < e.elts.length; ++i) {
+                item = this.vexpr(e.elts[i]);
+                if (allconsts && (item.indexOf('$const') == -1)) {
```
[the remark this answers is not in the document]
Comment (rixner) [reply src/compile.js]: I am definitely open to suggestions!

I added a comment, which doesn't fix the ugliness, but it at least explains it...

**Comment by rixner** [comment]
Is this languishing because people don't have time to think about it or because they share @meredydd 's unease?

**Comment by meredydd** [comment]
For what it's worth, my unease has been satisfied. I agree that, as immutable objects, it doesn't matter whether we give them separate object identities or not.

**Comment by bnmnetp** [comment]
@rixner I assume you are using this one in the wild and haven't encountered any issues with it, so I'm going to update and accept.  Since you and @meredydd are on board, and I have looked at it and don't see anything wrong with it.

**Comment by rixner** [comment]
@bnmnetp yes I have been using it in the wild for a while with no issues.

**Pull request merged**